cpp/src/arrow/compute/kernels/compare.cc

```diff
@@ -0,0 +1,184 @@
+      writer.Set();
+    } else {
+      writer.Clear();
+    }
+  }
+
+  writer.Finish();
+
+  return Status::OK();
+}
+
+template <typename ArrowType, CompareOperator Op>
+class CompareFunction final : public FilterFunction {
+  using ArrayType = typename TypeTraits<ArrowType>::ArrayType;
+  using ScalarType = typename TypeTraits<ArrowType>::ScalarType;
+
+ public:
+  explicit CompareFunction(FunctionContext* ctx) : ctx_(ctx) {}
+
+  Status Filter(const Array& input, const Scalar& scalar, ArrayData* output) const {
+    // Caller must cast
+    DCHECK(input.type()->Equals(scalar.type));
+    // Caller must not pass null scalar, use IsNull, IsNotNull
+    DCHECK(scalar.is_valid);
+    // Output must be a boolean array
+    DCHECK(output->type->Equals(boolean()));
+    // Output must be of same length
+    DCHECK_EQ(output->length, input.length());
+
+    auto input_data = input.data();
+
+    // Scalar is null, all comparisons are null.
+    if (!scalar.is_valid) return detail::SetAllNulls(ctx_, *input_data, output);
+
+    // Copy null_bitmap
+    RETURN_NOT_OK(detail::PropagateNulls(ctx_, *input_data, output));
+
+    uint8_t* bitmap_result = output->buffers[1]->mutable_data();
+    return CompareArrayScalar<ArrowType, Op>(static_cast<const ArrayType&>(input),
+                                             static_cast<const ScalarType&>(scalar),
+                                             bitmap_result);
+  }
+
+ private:
+  FunctionContext* ctx_;
+};
+
+template <typename ArrowType, CompareOperator Op>
+static inline std::shared_ptr<FilterFunction> MakeCompareFunctionTypeOp(
+    FunctionContext* ctx) {
+  return std::make_shared<CompareFunction<ArrowType, Op>>(ctx);
+}
+
+template <typename ArrowType>
+static inline std::shared_ptr<FilterFunction> MakeCompareFilterFunctionType(
+    FunctionContext* ctx, struct CompareOptions options) {
+  switch (options.op) {
+    case CompareOperator::EQUAL:
+      return MakeCompareFunctionTypeOp<ArrowType, CompareOperator::EQUAL>(ctx);
+    case CompareOperator::NOT_EQUAL:
+      return MakeCompareFunctionTypeOp<ArrowType, CompareOperator::NOT_EQUAL>(ctx);
+    case CompareOperator::GREATER:
+      return MakeCompareFunctionTypeOp<ArrowType, CompareOperator::GREATER>(ctx);
+    case CompareOperator::GREATER_EQUAL:
+      return MakeCompareFunctionTypeOp<ArrowType, CompareOperator::GREATER_EQUAL>(ctx);
+    case CompareOperator::LOWER:
+      return MakeCompareFunctionTypeOp<ArrowType, CompareOperator::LOWER>(ctx);
+    case CompareOperator::LOWER_EQUAL:
+      return MakeCompareFunctionTypeOp<ArrowType, CompareOperator::LOWER_EQUAL>(ctx);
+  }
+
+  return nullptr;
+}
+
+std::shared_ptr<FilterFunction> MakeCompareFilterFunction(FunctionContext* ctx,
+                                                          const DataType& type,
+                                                          struct CompareOptions options) {
+  switch (type.id()) {
+    case UInt8Type::type_id:
```

Comment:
Could use `VisitTypeInline`, though it's largely a matter of personal preference.